lib/ansible/parsing/dataloader.py

```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
         given = unquote(given)
         given = to_text(given, errors='surrogate_or_strict')
 
-        if given.startswith(u"/"):
-            return os.path.abspath(given)
-        elif given.startswith(u"~"):
-            return os.path.abspath(os.path.expanduser(given))
+        if RE_ABS_PATH.search(given):
+            path = unfrackpath(given)
         else:
             basedir = to_text(self._basedir, errors='surrogate_or_strict')
-            return os.path.abspath(os.path.join(basedir, given))
+            path = unfrackpath(os.path.join(basedir, given))
+
+        return path
```

Comment:
* The path_dwim() code was not doing the same thing as ```unfrackpath([...], follow=True)```.  The path_dwim() code was doing approximately the same things as ```unfrackpath([...],  follow=False)```.
* The type problem with the regex is that, on python2 it mixes a bytes regex with text data.  path_dwim() currently expects text strings so we need to use text everywhere to avoid mixing types.